crates/fuzzing/src/lib.rs

```diff
@@ -1,2 +1,176 @@
+//! Fuzzing infrastructure for Wasmtime.
+
+#![deny(missing_docs, missing_debug_implementations)]
+
 pub mod generators;
 pub mod oracles;
+
+use anyhow::{Context, Error};
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::process::{self, Command};
+use std::sync::{atomic, Once};
+
+/// Run a fuzz test on Wasm test case with automatic logging.
+///
+/// This is intended for defining the body of a `libfuzzer_sys::fuzz_target!`
+/// invocation.
+///
+/// Automatically prints out how to create a regression test that runs the exact
+/// same set of oracles.
+///
+/// It also binds the expression getting the wasm bytes to the variable, for
+/// example below the `wasm` variable is assigned the value
+/// `&my_test_case.as_wasm_bytes()`. This variable can be used within the body.
+///
+/// ```ignore
+/// use wasmtime_fuzzing::{oracles, with_log_wasm_test_case};
+///
+/// with_log_wasm_test_case!(&my_test_case.as_wasm_bytes(), |wasm| {
+///     oracles::compile(wasm);
+///     oracles::instantiate(wasm);
+/// });
+/// ```
+#[macro_export]
+macro_rules! with_log_wasm_test_case {
+    ( $wasm:expr , |$wasm_var:ident| $oracle:expr ) => {{
+        let $wasm_var = $wasm;
+        wasmtime_fuzzing::log_wasm_test_case(
+            &$wasm_var,
+            stringify!($wasm_var),
+            stringify!($oracle),
+        );
+        $oracle;
+    }};
+}
+
+/// Given that we are going to do a fuzz test of the given Wasm buffer, log the
+/// Wasm and its WAT disassembly, and preserve them to the filesystem so that if
+/// we panic or crash, we can easily inspect the test case.
+///
+/// This is intended to be used via the `with_log_wasm_test_case` macro.
+pub fn log_wasm_test_case(wasm: &[u8], wasm_var: &'static str, oracle_expr: &'static str) {
+    init_logging();
+
+    let wasm_path = wasm_test_case_path();
+    fs::write(&wasm_path, wasm)
+        .with_context(|| format!("Failed to write wasm to {:?}", &wasm_path))
+        .unwrap();
+    log::info!("Wrote Wasm test case to: {}", wasm_path.display());
+
+    let wat_path = wat_disassembly_path();
+    match wasm2wat(&wasm_path, &wat_path) {
+        Ok(wat) => {
+            log::info!("WAT disassembly:\n{}", wat);
+            log::info!(
+                "Also wrote test case's WAT disassembly to: {}",
+                wat_path.display()
+            );
+        }
+        Err(e) => {
+            log::info!("Failed to disassemble Wasm into WAT:\n{:?}", e);
+        }
+    };
+
+    log::info!(
+        "If this fuzz test fails, copy `{wasm_path}` to `wasmtime/crates/fuzzing/tests/regressions/my-fuzzing-regression.wasm` and add the following test to `wasmtime/crates/fuzzing/tests/regressions.rs`:
+
+```
+#[test]
+fn my_fuzzing_regression_test() {{
+    let {wasm_var} = include_bytes!(\"./regressions/my-fuzzing-regression.wasm\");
```

Comment:
Ideally our regression tests would be `*.wat` files, so would it be possible to generate tests that use the `*.wat` file instead?

Reply:
Yeah, there are possibly cases where we panic on invalid wasm or something, but for most test cases we should definitely be able to do this.